internal/emigration/emigrate-nobles.lua

```diff
@@ -0,0 +1,352 @@
+    end
+
+    -- erase the unit from the fortress entity
+    for k,v in ipairs(oldSite.histfig_ids) do
+        if v == histFigId then
+            df.global.plotinfo.main.fortress_entity.histfig_ids:erase(k)
+            break
+        end
+    end
+    for k,v in ipairs(oldSite.hist_figures) do
+        if v.id == histFigId then
+            df.global.plotinfo.main.fortress_entity.hist_figures:erase(k)
+            break
+        end
+    end
+    for k,v in ipairs(oldSite.nemesis) do
+        if v.figure.id == histFigId then
+            df.global.plotinfo.main.fortress_entity.nemesis:erase(k)
+            df.global.plotinfo.main.fortress_entity.nemesis_ids:erase(k)
+            break
+        end
+    end
+
+    -- remove the old entity link and create new one to indicate former membership
+    histFig.entity_links:insert("#", {new = df.histfig_entity_link_former_memberst, entity_id = oldSite.id, link_strength = 100})
+    for k,v in ipairs(histFig.entity_links) do
+        if v._type == df.histfig_entity_link_memberst and v.entity_id == oldSite.id then
+            histFig.entity_links:erase(k)
+            break
+        end
+    end
+end
+
+-- adapted from emigration::desert()
+---@param unit df.unit
+---@param toSite df.world_site
+local function emigrate(unit, toSite)
+    local histFig = df.historical_figure.find(unit.hist_figure_id)
+    if histFig == nil then qerror("could not find histfig!") end
+
+    local fortEnt = df.global.plotinfo.main.fortress_entity
+
+    -- mark for leaving
+    unit.following = nil
+    unit.civ_id = playerCiv.id -- should be redundant but oh well
+    unit.flags1.forest = true
+    unit.flags2.visitor = true
+    unit.animal.leave_countdown = 2
+
+    -- remove current job
+    if unit.job.current_job then dfhack.job.removeJob(unit.job.current_job) end
+
+    -- break up any social activities
+    for _, actId in ipairs(unit.social_activities) do
+        local act = df.activity_entry.find(actId)
+        if act then act.events[0].flags.dismissed = true end
+    end
+
+    removeUnitFromSiteEntity(unit, histFig, fortEnt)
+    local siteGov = addHistFigToSite(histFig, toSite)
+
+    -- announce the changes
+    local unitName = dfhack.df2console(dfhack.units.getReadableName(unit))
+    local siteName = dfhack.df2console(dfhack.translation.translateName(toSite.name, true))
+    local govName = dfhack.df2console(dfhack.translation.translateName(siteGov.name, true))
+    local line = unitName .. " has left to join " ..govName.. " as lord of " .. siteName .. "."
+    print("[+] "..dfhack.df2console(line))
+    dfhack.gui.showAnnouncement(line, COLOR_WHITE)
+end
+
+---@param unit df.unit
+local function inStrangeMood(unit)
+    local job = unit.job.current_job
+    if not job then return false end
+
+    local jobType = job.job_type -- taken from notifications::for_moody()
+    return df.job_type_class[df.job_type.attrs[jobType].type] == 'StrangeMood'
```

Comment:
in general, you should not be canceling any job with the `special` flag. Attempting to cancel a special job may crash the game